tree/dataframe/src/RNTupleDS.cxx

```diff
@@ -460,6 +462,41 @@ bool RNTupleDS::SetEntry(unsigned int, ULong64_t)
    return true;
 }
 
+void RNTupleDS::ExecStaging()
+{
+   while (true) {
+      std::unique_lock lock(fMutexStaging);
+      fCvStaging.wait(lock, [this] { return fIsReadyForStaging || fStagingThreadShouldTerminate; });
+      if (fStagingThreadShouldTerminate)
+         return;
+
+      assert(!fHasNextSources);
+      StageNextSources();
```

Comment:
Why are we processing `fNSlots` files as a single atomic operation? Why not notify the main thread every time we load a new file?

Reply:
This is related to the way `GetEntryRanges()` currently works. The call returns the next batch of ranges for all slots together. So we would still need to wait until all `fNSlots` next files are opened. The only thing we could already do is to process the metadata (i.e., deserialize and interpret the descriptor). But it doesn't seem to be worthwhile the additional code.

Eventually, I would like to see an RDataFrame interface that asks for new work whenever a slot finishes a task (a batch of events). That requires bigger changes in RDF though.